File: solutions/javascript/algorithms/algorithm_bubble_sort.js

// create the array to sort
var arr = createArrayToSort(1000, 0, 10000)

// create a sorted version of the array for validation
var validationArr = [].concat(arr).sort(function (a, b){ return a - b })

// sort the array
var sortedArr = bubbleSort([].concat(arr))

// compare the native and custom sorts
compareArrays(validationArr, sortedArr)

// done
console.log("THE ARRAYS MATCHED!")

/**
 * Sort an array w/ bubble sort
 * @param  {Array.<number>} sortArr the array to sort
 * @return {Array.<number>} returns the array to be sorted
 */
function bubbleSort(sortArr) {
  var swapped
  do {
    swapped = false
    for (var i = 0; i < sortArr.length - 1; i++) {
      if (sortArr[i] > sortArr[i + 1]) {
        swapped = true
        var temp = sortArr[i]
        sortArr[i] = sortArr[i + 1]
        sortArr[i + 1] = temp
      }
    }
  } while (swapped)

  return sortArr
}

/**
 * Compare 2 arrays and make sure each element matches
 *
 * @param  {Array.<number>} arr1 the first array
 * @param  {Array.<number>} arr2 the second array
 */
function compareArrays(arr1, arr2) {
  if (arr1.length != arr2.length) throw new Error("The array lengths did not match")
  for (var i = 0; i < arr1.length; i++) {
    if (arr1[i] != arr2[i]) throw new Error("The arrays did not match")
  }
}

/**
 * Create an array of numbers to sort
 *
 * @param  {number} length the number of items to insert into the array
 * @param  {number} min    the minimium value for each item
 * @param  {number} max    the maximum value for each item (non-inclusive)
 * @return {Array.<number>} The array of numbers
 */
function createArrayToSort(length, min, max) {
  var rangeSize = max - min
  var items = []
  for (var i = 0; i < length; i++) {
    items.push(Math.floor(Math.random() * rangeSize) + min)
  }
  return items
}

File: solutions/algorithms/algorithm_bubble_sort.py
import random

def bubbleSort(arr):
  newArr = arr[:]
  listLen = len(arr)
  swapped = True
  while swapped:
    swapped = False
    for i in range(0, listLen - 1):
      if newArr[i] > newArr[i + 1]:
        temp = newArr[i]
        newArr[i] = newArr[i + 1]
        newArr[i + 1] = temp
        swapped = True
  return newArr

def createArrayToSort(len, min, max):
  arr = []
  for i in range(0, len):
    arr.append(random.randint(min, max))
  return arr

def compareArrays(arr1, arr2):
  len1 = len(arr1)
  len2 = len(arr2)
  if len1 != len2:
    print "The list lengths did not match"
    return False
  for i in range(0, len1):
    if arr1[i] != arr2[i]:
      print "The lists did not match at index " + i
      return False
  return True

arr = createArrayToSort(1000, 0, 10000)
validationArr = sorted(arr)
sortedArr = bubbleSort(arr)

listsMatch = compareArrays(validationArr, sortedArr)
if listsMatch:
  print "THE LISTS MATCH"